### ISSUE TITLE
imshow longdouble not truly supported

### ISSUE THREAD

**Issue by QuLogic** [matplotlib/matplotlib#10342]
### Bug report

Support for `longdouble` was theoretically added in #8447. There is also code to test it out; unfortunately, it doesn't actually attempt to draw anything. Adding the draw causes it to crash.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np

fig, ax = plt.subplots()
ax.imshow(np.zeros((3, 3), dtype=np.longdouble))

fig.canvas.draw()
```

**Actual outcome**

```pytb
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/elliott/code/matplotlib/lib/matplotlib/backends/backend_tkagg.py", line 299, in draw
    FigureCanvasAgg.draw(self)
  File "/home/elliott/code/matplotlib/lib/matplotlib/backends/backend_agg.py", line 432, in draw
    self.figure.draw(self.renderer)
  File "/home/elliott/code/matplotlib/lib/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/elliott/code/matplotlib/lib/matplotlib/figure.py", line 1337, in draw
    renderer, self, artists, self.suppressComposite)
  File "/home/elliott/code/matplotlib/lib/matplotlib/image.py", line 138, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/elliott/code/matplotlib/lib/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/elliott/code/matplotlib/lib/matplotlib/axes/_base.py", line 2560, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/home/elliott/code/matplotlib/lib/matplotlib/image.py", line 138, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/elliott/code/matplotlib/lib/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/elliott/code/matplotlib/lib/matplotlib/image.py", line 568, in draw
    renderer, renderer.get_image_magnification())
  File "/home/elliott/code/matplotlib/lib/matplotlib/image.py", line 795, in make_image
    unsampled=unsampled)
  File "/home/elliott/code/matplotlib/lib/matplotlib/image.py", line 413, in _make_image
    self.get_filterrad() or 0.0)
ValueError: Unsupported dtype
```

**Matplotlib version**
  * Operating system: Fedora
  * Matplotlib version: 2.1.0 & `master`
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.6
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by tacaswell** [comment]
Tagging as a 3.0 problem because we have not had any bug reports about this not working.

**Comment by QuLogic** [comment]
Well, there was the original #8445.

**Comment by efiring** [comment]
Looks easy.  The block starting on line 376 of image.py just needs to catch longdouble and set scaled_dtype to np.float64.

**Comment by b52** [comment]
Not sure if this is related, but I have a similar error resulting at least in a very similar stack trace:

```
  File "/usr/lib/python3.6/site-packages/matplotlib/pyplot.py", line 697, in savefig
    res = fig.savefig(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/matplotlib/figure.py", line 1573, in savefig
    self.canvas.print_figure(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/matplotlib/backend_bases.py", line 2252, in print_figure
    **kwargs)
  File "/usr/lib/python3.6/site-packages/matplotlib/backends/backend_agg.py", line 593, in print_jpg
    buf, size = self.print_to_buffer()
  File "/usr/lib/python3.6/site-packages/matplotlib/backends/backend_agg.py", line 563, in print_to_buffer
    FigureCanvasAgg.draw(self)
  File "/usr/lib/python3.6/site-packages/matplotlib/backends/backend_agg.py", line 464, in draw
    self.figure.draw(self.renderer)
  File "/usr/lib/python3.6/site-packages/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/matplotlib/figure.py", line 1144, in draw
    renderer, self, dsu, self.suppressComposite)
  File "/usr/lib/python3.6/site-packages/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File "/usr/lib/python3.6/site-packages/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 2426, in draw
    mimage._draw_list_compositing_images(renderer, self, dsu)
  File "/usr/lib/python3.6/site-packages/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File "/usr/lib/python3.6/site-packages/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/matplotlib/image.py", line 543, in draw
    renderer, renderer.get_image_magnification())
  File "/usr/lib/python3.6/site-packages/matplotlib/image.py", line 770, in make_image
    unsampled=unsampled)
  File "/usr/lib/python3.6/site-packages/matplotlib/image.py", line 426, in _make_image
    self.get_filternorm() or 0.0, self.get_filterrad() or 0.0)
ValueError: 3-dimensional arrays must be of dtype unsigned byte, unsigned short, float32 or float64
```

I basically created various subplots with `np.float16` and `np.float64` 2D images without color channel, i.e., `ndim == 2`. I assume the `np.float16` is the issue, because when I convert it to a `np.float32` everything is fine.

**Comment by dstansby** [comment]
Fixed by #10768
